klangbecken.liq

```diff
@@ -56,23 +56,80 @@ ALSA_DEVICE = getenv_fallback("KLANGBECKEN_ALSA_DEVICE", "default")
 # PLAYLISTS                                         #
 # ================================================= #
 
+# parse index.json and keep it up to date
+index = ref []
+index_filename = path.concat(DATA_DIR, "index.json")
+def read_index() =
+  log("updating index", label="update_index")
+  index := of_json(default=[("list", [("key", "value")])], file.contents(index_filename))
+end
+if file.exists(index_filename) then read_index() end
+file.watch(index_filename, read_index)
+
+# calculate waiting time for repeating a track depending on its playlist
+def calc_wait(playlist) =
+  if playlist == "music" then 259200 # 3 days
+  elsif playlist == "classics" then 86400 # 1 day
+  elsif playlist == "jingles" then 3600 # 1 hour
+  else 0 end
+end
+
+# check if track was played recently
+skipped = ref 0
+def check_next_func(r) = 
+  # get id from filename
+  id = basename(request.filename(r))
+  id = string.replace(pattern="\..*", fun(_) -> "", id)
+  # get info about track in index
+  meta = list.assoc(default=[], id, !index)
+  # get last_play as a unix timestamp
+  last_play = meta["last_play"]
+
+  if string.length(last_play) == 0 then
+    log("track was never played before: #{request.filename(r)}", label="check_next_func")
+    true
+  else
+    last_play_unix = int_of_string(string.trim(get_process_output("date --date=#{last_play} +%s")))
+    now_unix = int_of_float(gettimeofday())
+    diff = now_unix - last_play_unix
+    if diff < calc_wait(meta["playlist"]) then
+      # some failsafe here would be nice
+      skipped := !skipped + 1
+      if !skipped >= 10 then
+        skipped := 0
+        log("too many skipped tracks, playing #{request.filename(r)} anyway", label="check_next_func")
```

Comment:
maybe we can still add the diff time to the log

Reply:
Yeah that's a good idea. Will do this.